add UI for editing website collection entries

aliceriot opened the pull request:
#### Pre-Flight checklist

- [x] Screenshots and design review for any changes that affect layout or styling
  - [x] Desktop screenshots
  - [ ] Mobile width screenshots
- [x] Testing
  - [x] Code is tested
  - [x] Changes have been manually tested


#### What are the relevant tickets?

closes #376
closes #397

#### What's this PR do?

This adds UI to the website collection editing interface which allows you to add websites to a website collection and also to re-order existing websites in the collection.

#### How should this be manually tested?

Go to `/collections` and check out a collection you have. There should be a new section in the edit form with a field for picking a website. If you pick one and click 'Add' a request should go off to the backend, and then it should appear in your list. Once you've added a few websites to your list you should be able to re-order them as you please.


#### Screenshots (if appropriate)

<img width="577" alt="Screen Shot 2021-07-15 at 2 59 01 PM" src="https://user-images.githubusercontent.com/6207644/125842621-050c6648-307f-494c-b62c-92a84255af88.png">
<img width="679" alt="Screen Shot 2021-07-15 at 2 59 25 PM" src="https://user-images.githubusercontent.com/6207644/125842623-a2049377-43a1-4999-8800-2e1b61ebf439.png">
<img width="628" alt="Screen Shot 2021-07-15 at 2 59 20 PM" src="https://user-images.githubusercontent.com/6207644/125842625-21857fe7-b9f2-41cb-b161-e3a78558cd3c.png">
<img width="657" alt="Screen Shot 2021-07-15 at 2 59 06 PM" src="https://user-images.githubusercontent.com/6207644/125842627-b7dc3e70-cef9-4eba-8fe0-13dfb938842c.png">

gsidebo commented on static/js/components/WebsiteCollectionItemsEditor.tsx:
```diff
@@ -0,0 +1,112 @@
+
+  return (
+    <div className="collection-item-editor pt-5 pb-3">
+      <h4>Courses in this collection</h4>
```
One quick thing I noticed

```suggestion
      <h4>Sites in this collection</h4>
```

aliceriot replied:
@Ferdi wondering your thoughts on this, I made this header 'Courses in this collection' following what was in the mockups, should we stick with that or go with the more general 'Sites in this collection' ?

noisecapella commented on static/js/hooks/useAllWebsites.ts:
```diff
@@ -0,0 +1,48 @@
+import { WEBSITE_CONTENT_PAGE_SIZE } from "../constants"
+
+/**
+ * A rather brute-force way to fetch all the websites, to be able to
```
This is fetching more than 150 requests on my local machine. Is there a way to pass in (or just hardcode) a larger value for `limit`?

aliceriot replied:
ahh yeah I don't have a realistic set of courses locally I think, haha

I'll look into this stuff tomorrow

noisecapella commented on websites/views.py:
```diff
@@ -420,7 +420,7 @@ class WebsiteCollectionItemViewSet(NestedViewSetMixin, viewsets.ModelViewSet):
 
     permission_classes = (HasWebsiteCollectionItemPermission,)
-    pagination_class = DefaultPagination
+    # pagination_class = DefaultPagination
```
Stray comment

noisecapella commented:
I don't think this is workable as is even with a higher limit on websites fetched because it just generates too many requests and stores too much data in the DOM. How complicated would it be to use the async functionality of react-select? https://react-select.com/home#async I guess that would require an extra filter on the backend to support typeahead functionality.

noisecapella commented:
Drag and drop works but it feels clunky. The item text also gains focus:
![Screenshot from 2021-07-15 17-21-18](https://user-images.githubusercontent.com/863262/125859370-23b9e22d-737c-42c1-a021-39f106f9b81a.png)

I feel like maybe there is some CSS conflict which is preventing some of the drag and drop behavior but that's just a random guess. The [examples](https://5fc05e08a4a65d0021ae0bf2-cbfrwoklyt.chromatic.com/?path=/story/presets-sortable-vertical--basic-setup) show an animation which doesn't appear in this PR

noisecapella commented:
If you add the same course twice it causes a react warning about duplicate keys

aliceriot commented:
- I have some work in progress which will prevent adding duplicate courses
- for the async thing, I can look into that tomorrow to see if it's relatively straightforward
- what browser are you using? I don't see that in up to date Chrome on macOS

noisecapella commented:
I saw the drag and drop behavior on both Chrome and Firefox, but I'm happy to push it off to another PR since it works except for the animation

aliceriot commented:
@noisecapella ok two things:

- I added a quick implementation of a typeahead sort of thing, basically just added an optional 'search' query param to the website listing API and added code to use the AsyncSelect component from react-select with it.
- also realized I had accidentally deleted the CSS transition stuff from the draggable component! so that was what you were noticing I think. it should be fixed now.

no tests yet, but w/ the async thing lmk how you think it works.

noisecapella commented:
I'm seeing this error when I click on a collection:
![Screenshot from 2021-07-16 11-38-53](https://user-images.githubusercontent.com/863262/125973290-a14825c5-2cec-4b22-8723-162d07376ce1.png)

aliceriot commented:
Ahh I think I need to add a default value, is that on a collection which
doesn’t have any items in it yet?

On Fri, Jul 16, 2021 at 11:39 AM George Schneeloch ***@***.***>
wrote:

> I'm seeing this error when I click on a collection:
> [image: Screenshot from 2021-07-16 11-38-53]
> <https://user-images.githubusercontent.com/863262/125973290-a14825c5-2cec-4b22-8723-162d07376ce1.png>
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mitodl/ocw-studio/pull/391#issuecomment-881539562>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABPLRHCYOPAC746L2BTBQMTTYBHCXANCNFSM5AOEGVZA>
> .
>
-- 
Alice

noisecapella commented:
Yes, but looking at the code it seems like `items` is the response from the website, not the results from that list. So you might want to do `items.results.map` or make a similar change elsewhere

noisecapella commented:
I think the root of that bug is the `websiteCollectionItemsRequest.transform` function. The `body` is a response object, not a `WebsiteCollectionItem[]`. If I change `[id]: body` to `[id]: body.results` in that function the problem goes away

noisecapella commented:
The typeahead functionality works great. It would be nice to have some initial list of courses before you type something but I don't know what should go there. For now can we add placeholder text like `Search courses...` when it's blank?

I see some visual glitches with the drag and drop but otherwise it works great. We can file an issue and deal with that later: 
![Screenshot from 2021-07-16 13-18-47](https://user-images.githubusercontent.com/863262/125985611-c12ccaa3-5cbb-4f00-a7a9-0a568121a9cd.png)
![Screenshot from 2021-07-16 13-19-11](https://user-images.githubusercontent.com/863262/125985612-98ddacdd-b02d-4963-a76d-85963171c225.png)

aliceriot commented:
> I think the root of that bug is the websiteCollectionItemsRequest.transform function. The body is a response object, not a WebsiteCollectionItem[]. If I change [id]: body to [id]: body.results in that function the problem goes away

ohh this was because I had the pagination commented out accidentally on the WebsiteCollectionItem viewset

aliceriot commented:
<img width="581" alt="Screen Shot 2021-07-16 at 3 30 12 PM" src="https://user-images.githubusercontent.com/6207644/125999441-8c8b77ae-9e3e-40d9-9587-25bfaa3f1daf.png">

I went with this as a placeholder based on some copy in the mockups

noisecapella commented:
Is there an argument to delay requests to the API to reduce the number of requests made while typing? If it's not easy to do we can file a ticket for it

noisecapella commented:
The results are a little confusing since multiple courses will have the same title. I'll file an issue to address in a future PR
![Screenshot from 2021-07-16 15-56-09](https://user-images.githubusercontent.com/863262/126002069-8ea6e90d-d462-4032-a50f-5f1391fbdb82.png)

noisecapella commented on static/js/components/forms/WebsiteCollectionItemForm.tsx:
```diff
@@ -0,0 +1,74 @@
+
+  return (
+    // @ts-ignore
+    <Formik onSubmit={onSubmitCB} initialValues={{ website: undefined }}>
```
What does the `website: undefined` do here?

aliceriot replied:
sort of a weird thing, I'm going to probably add a comment to clarify

basically Formik and react-select disagree here! Formik thinks that `initialValues` should have `""`, but react-select does not consider `""` to be an empty value (afaict). So before I was passing `{website: "" }` as the `initialValues` prop, but this was causing the placeholder not to show up in react-select. Once I changed it to `undefined` the react-select placeholder started to show up. The `@ts-ignore` is here because Formik wants this prop to be `Record<string, string>` I think (or something along those lines).

noisecapella commented on static/js/query-configs/website_collections.ts:
```diff
@@ -136,3 +150,94 @@ export const createWebsiteCollectionMutation = (
+      newArray[oldIndex] = next
+
+      return {
+        [collectionId]: arrayMove(newArray, oldIndex, next.position)
```
Do we need to adjust the `position` of any of the other items?

aliceriot replied:
hmm we could if we wanted to make sure we have full consistency on the frontend, but I don't think it's strictly necessary. We never read, on the frontend, from the `position` attribute to determine what order to display things, and the logic here and in the `handleDrag` function in `WebsiteCollectionItemsEditor` works based on the position of the items in the array rather than the `position` attribute. Another way to say that, is that we're only really using the `position` attribute when we're doing a `PATCH` on one of our existing records.
